src/tls/socket-activation-helper.c

```diff
@@ -0,0 +1,193 @@
+} instance_types[] = {
+  {"https.sock", {"--for-tls-proxy", "--port", "0"}},
+  {"http-redirect.sock", {"--proxy-tls-redirect", "--no-tls", "--port", "0"}},
+  {"http.sock", {"--no-tls", "--port", "0"}},
+};
+
+static int socket_to_pid[N_ELEMENTS (instance_types)];
+static struct pollfd ufds[N_ELEMENTS (instance_types)];
+
+static void
+handle_cockpit_ws_exited (int sigid)
+{
+  pid_t pid;
+  while ((pid = waitpid (-1, NULL, WNOHANG)) > 0)
+    {
+      for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+        {
+          if (pid == socket_to_pid[i])
+            {
+              ufds[i].events = POLLIN;
+              socket_to_pid[i] = 0;
+              return;
+            }
+        }
+    }
+  /* This can only be reached in case cockpit-ws exited fast and the parent
+   * process did not manage to insert its PID into the array
+   */
+  errx (1, "Could not find the process in socket_to_pid array");
+}
+
+static pid_t
+spawn_cockpit_ws (const char *ws_path, int fd,
+                  struct instance_type sock_instance)
+{
+  pid_t pid = fork ();
+
+  if (pid == -1)
+    {
+      err (1, "fork() failed");
+    }
+  else if (pid > 0)
+    {
+      return pid;
+    }
+  else
+    {
+      const char *args[MAX_COCKPIT_WS_ARGS + 1];
+      int i = 0;
+      static char listen_pids[20];
+
+      /* pass the socket to ws like systemd activation does, see sd_listen_fds(3) */
+      if (dup2 (fd, SD_LISTEN_FDS_START) < 0)
+        err (1, "dup2() failed");
+
+      if (snprintf (listen_pids, sizeof (listen_pids), "LISTEN_PIDS=%i",
+                    getpid ()) >= sizeof listen_pids)
+        errx (1, "asprintf() failed");
+
+      char *env[] = {
+        "LISTEN_FDS=1", listen_pids, "SD_LISTEN_FDS_START=3", (char *) 0
+      };
+
+      args[0] = ws_path;
+      for (i = 0; sock_instance.argv[i] != NULL; i++)
+        args[i + 1] = sock_instance.argv[i];
+      args[i + 1] = NULL;
+
+      if (execve (ws_path, (char **) args, env) < 0)
+        err (1, "spawning cockpit-ws instance failed");
+
+      exit (0);
+    }
+}
+
+int
+main (int argc, char *argv[])
+{
+  if (argc != 3)
+    errx (1, "Usage: socket-activation-helper $WS_PATH $SOCKETS_DIR");
+
+  const char *ws_path = argv[1];
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    ufds[i].fd = -1;
+
+  /* Set up signal handler for when cockpit-ws has exited */
+  struct sigaction sa;
+  sigemptyset (&sa.sa_mask);
+  sa.sa_flags = 0;
+  sa.sa_handler = handle_cockpit_ws_exited;
+  sigaction (SIGCHLD, &sa, NULL);
+
+  for (int i = 0; i < N_ELEMENTS (instance_types); i++)
+    {
+      struct sockaddr_un addr = {
+        .sun_family = AF_UNIX
+      };
+      int listen_fd;
+      int rv;
+
+      socket_to_pid[i] = 0;
+
+      rv = snprintf (addr.sun_path, sizeof (addr.sun_path), "%s/%s", argv[2],
+                     instance_types[i].sockname);
+      if (rv <= 0 || rv >= sizeof addr.sun_path)
+        errx (1, "snprintf() failed");
+
+      if (unlink (addr.sun_path) < 0)
```

Comment:
It's OK if `unlink()` fails with `ENOENT`, as is the expected case on first run (when the socket doesn't exist).